Depend on two versions of maia

klochowicz opened the pull request:
With the advent of libp2p-based protocols, we have the opportunity to clearly
demarcate older and newer takers, making it possible to roll out recent maia bugfixes.

After this change:
- libp2p-based protocols will use maia v0.2.0
- legacy networking will use maia v0.1.0 (`maia_deprecated`), which matches maia
    version shipped with legacy takers

The contents of setup_contract.rs (protocol logic) has been copied wholesale to
`setup_contract_deprecated.rs` with the expectation that the whole file will be
deleted when we decide to drop support for takers using legacy networking layer.

Co-authored-by: Philipp Hoenisch <philipp@hoenisch.at>

bonomat committed:
Depend on two versions of maia

With the advent of libp2p-based protocols, we have the opportunity to clearly
demarcate older and newer takers, making it possible to roll out recent maia bugfixes.

After this change:
- libp2p-based protocols will use maia v0.2.0
- legacy networking will use maia v0.1.0 (`maia_deprecated`), which matches maia
    version shipped with legacy takers

The contents of setup_contract.rs (protocol logic) has been copied wholesale to
`setup_contract_deprecated.rs` with the expectation that the whole file will be
deleted when we decide to drop support for takers using legacy networking layer.

Co-authored-by: Philipp Hoenisch <philipp@hoenisch.at>

klochowicz commented on Cargo.toml:
```diff
@@ -21,6 +21,7 @@ resolver = "2"
 xtra = { git = "https://github.com/comit-network/xtra" } # We need to use unreleased patches.
-maia = { git = "https://github.com/comit-network/maia" } # Unreleased
-maia-core = { git = "https://github.com/comit-network/maia" } # Unreleased
+maia = { git = "https://github.com/comit-network/maia", rev = "cf41b907e6eb68245dcb7c561ec7d4e6bcc4ed19" } # FIXME: point back to master after https://github.com/comit-network/maia/pull/36 gets merged
```
this will need to be adjusted after maia changes get upstreamed.

luckysori committed:
Ensure that supervised actor stays alive

Other than the restart functionality provided by the
`supervisor::Actor`, by storing a strong `xtra::Address` to the actor
being supervised inside of `supervisor::Actor` we ensure that the
supervised actor can never be shut down due to a lack of strong
addresses existing.

luckysori committed:
Stop storing xtra::Addresses redundantly

We had several instances where we were storing an `xtra::Address` for
supervised actors in order to ensure that they were not shut down as
soon as all of their addresses had been dropped.

Now that the `supervisor::Actor` does this for you, we don't need to
do so.

luckysori committed:
Simplify supervised listener actor constructor

luckysori committed:
Make supervisor::UnitReason public

So that we can use the simple `new` constructor on
`supervisor::Actor`.

We must implement `Clone` and `Copy` too so that the we don't trip on
the lint `missing-copy-implementations`.

luckysori committed:
Supervise the maker's ping actor

This serves two overlapping purposes:

1. Keep the `ping::Actor` alive by not dropping its only strong
`xtra::Address`.
2. Restart the `ping::Actor` if it is ever stopped or panics.

Point 1 therefore is a fix for
https://github.com/itchysats/itchysats/issues/2103.

da-kami commented on Cargo.toml:
```diff
@@ -21,6 +21,7 @@ resolver = "2"
-maia-core = { git = "https://github.com/comit-network/maia" } # Unreleased
+maia = { git = "https://github.com/comit-network/maia", rev = "cf41b907e6eb68245dcb7c561ec7d4e6bcc4ed19" } # FIXME: point back to master after https://github.com/comit-network/maia/pull/36 gets merged
+
+maia-core = { git = "https://github.com/comit-network/maia", rev = "e419cb2cf19bed6ec53eaa3672408a6205a9b705", package = "maia-core" } # Unreleased version, pinned before maia's breaking change
```
🙃 `Unreleased version` this will get outdated as soon as we release. Not sure this helps. 

We can assume that `0.4.19` will be the release. Maybe just hardcode the version here...?

klochowicz replied:
the "Unreleased" comment refers to `maia`, not itchysats! (and it was there before, I just extended it)

da-kami reviewed (commented):
I am a bit confused by the change. 
Where in the legacy code is the decision being made if the old or the new `contract_setup` is being called? How does the maker distinguish between the old and the new code?

klochowicz committed:
Use deprecated maia version for calculating collab close transaction

For the time being, use deprecated maia version in collab close unconditionally.
Upcoming libp2p-based collab settlement protocol should use new maia version
when its implemented.

Note that we still use new `maia` versions in `model` crate:
- maia::finalize_spend_transaction
- maia::lock_descriptor
- maia::spending_tx_sighash

This is OK, as they do not have any business logic in them and are trivial to inspect.
In hindsight, these functions should be located in `maia_core`, but they cannot
be migrated now as maia bugfix (that made it backwards-incompatible) is already merged.

luckysori commented on model/src/cfd.rs:
```diff
@@ -1866,7 +1866,7 @@ impl Dlc {
             (outpoint, amount)
         };
-        let (tx, sighash) = maia::close_transaction(
+        let (tx, sighash) = maia_deprecated::close_transaction(
```
❓ Can you remind me why we want to use the buggy version unconditionally? And maybe add the explanation explicitly to the commit message if possible.

klochowicz replied:
we are not rolling out collab close on libp2p just yet, and this code has changed quite a bit on the branch due to the fact that we've been redesigning the protocol so that both parties publish.
If we end up using the same function, then we'll need to add a condition here, yes. I figured it would be easier to assess on the collab close branch (and for now the crucial bit is preserving the backwards compatibility).

I will put this in the comment here and also in the commit message

luckysori commented on Cargo.toml:
```diff
@@ -21,6 +21,7 @@ resolver = "2"
-maia = { git = "https://github.com/comit-network/maia" } # Unreleased
-maia-core = { git = "https://github.com/comit-network/maia" } # Unreleased
+maia = { git = "https://github.com/comit-network/maia", rev = "cf41b907e6eb68245dcb7c561ec7d4e6bcc4ed19" } # FIXME: point back to master after https://github.com/comit-network/maia/pull/36 gets merged
+
```
🔧 Unwanted newline?

luckysori reviewed (approved):
LGTM.

When you fix the tests, could you apply the changes to the first commit?
